src/regex_impl.rs

```diff
@@ -492,6 +495,112 @@ impl<W: CodeUnitWidth> Regex<W> {
+    ) -> Result<Cow<'s, [W::SubjectChar]>, Error>
+    where
+        [<W as CodeUnitWidth>::PCRE2_CHAR]: ToOwned,
+        W::PCRE2_CHAR: TryInto<W::SubjectChar>,
+        <<W as CodeUnitWidth>::PCRE2_CHAR as TryInto<<W as CodeUnitWidth>::SubjectChar>>::Error:
+            std::fmt::Debug,
+    {
+        self.replace_impl(subject, replacement, false, extended)
+    }
+
+    pub fn replace_all<'s>(
+        &self,
+        subject: &'s [W::SubjectChar],
+        replacement: &'s [W::SubjectChar],
+        extended: bool,
+    ) -> Result<Cow<'s, [W::SubjectChar]>, Error>
+    where
+        [<W as CodeUnitWidth>::PCRE2_CHAR]: ToOwned,
+        W::PCRE2_CHAR: TryInto<W::SubjectChar>,
+        <<W as CodeUnitWidth>::PCRE2_CHAR as TryInto<<W as CodeUnitWidth>::SubjectChar>>::Error:
+            std::fmt::Debug,
+    {
+        self.replace_impl(subject, replacement, true, extended)
+    }
+
+    #[inline]
+    fn replace_impl<'s>(
+        &self,
+        subject: &'s [W::SubjectChar],
+        replacement: &'s [W::SubjectChar],
+        replace_all: bool,
+        extended: bool,
+    ) -> Result<Cow<'s, [W::SubjectChar]>, Error>
+    where
+        [<W as CodeUnitWidth>::PCRE2_CHAR]: ToOwned,
+        W::PCRE2_CHAR: TryInto<W::SubjectChar>,
+        <<W as CodeUnitWidth>::PCRE2_CHAR as TryInto<<W as CodeUnitWidth>::SubjectChar>>::Error:
+            std::fmt::Debug,
+    {
+        let mut options: u32 = 0;
+        options |= PCRE2_SUBSTITUTE_OVERFLOW_LENGTH;
+        // TODO: this should probably be configurabe from user-side
+        options |= PCRE2_SUBSTITUTE_UNSET_EMPTY;
+        if extended {
+            options |= PCRE2_SUBSTITUTE_EXTENDED;
+        }
+        if replace_all {
+            options |= PCRE2_SUBSTITUTE_GLOBAL;
+        }
+
+        // TODO: we can use MaybeUninit to avoid allocation
+        let mut capacity = 256;
+        let mut output: Vec<W::PCRE2_CHAR> = Vec::with_capacity(capacity);
+        capacity = output.capacity();
+
+        let mut rc = unsafe {
+            self.code
+                .substitute(subject, replacement, 0, options, &mut output, &mut capacity)
+        };
+
+        if let Err(e) = &rc {
+            if e.code() == PCRE2_ERROR_NOMEMORY {
+                if output.try_reserve(capacity - output.capacity()).is_err() {
+                    return Err(rc.unwrap_err());
+                }
+                capacity = output.capacity();
+                rc = unsafe {
+                    self.code.substitute(
+                        subject,
+                        replacement,
+                        0,
+                        options,
+                        &mut output,
+                        &mut capacity,
+                    )
+                };
+            }
+        }
+
+        let s = match rc? {
+            0 => Cow::Borrowed(subject),
+            _ => {
+                // +1 to account for null terminator
+                unsafe { output.set_len(capacity + 1) }; 
```

Comment:
I don't think this `set_len` should be necessary? We use `take(capacity)` down below so the length should not matter.

Reply:
Rust thinks the vector has a length of zero before we set the length, so we would always end up with an empty array (since the iterator never gives any elements)  in that case

Reply:
Got it - this avoids zeroing the Vec.